File: server/internal/serverstore/departments.go
package serverstore

import (
	"database/sql"
	"errors"
)

// ciColumnCmp returns a case-insensitive equality comparison between two
// columns (not a bound parameter). Unlike CaseInsensitiveCmp, both sides are
// column references (e.g. sg.grantee = g.name inside a correlated subquery),
// so there is no `?` placeholder. SQLite uses COLLATE NOCASE; PG folds both
// sides with LOWER.
func ciColumnCmp(colA, colB string) string {
	return "LOWER(" + colA + ") = LOWER(" + colB + ")"
}

// Group is an organization unit: a department with optional parent
// (pyramid) and leader (department head).
type Group struct {
	ID          int64
	Name        string
	ParentID    int64
	LeaderID    int64
	Description string
}

// DepartmentInfo is the admin-view shape of a department.
type DepartmentInfo struct {
	ID          int64  `json:"id"`
	Name        string `json:"name"`
	ParentID    int64  `json:"parent_id"`
	LeaderID    int64  `json:"leader_id"`
	LeaderName  string `json:"leader_name,omitempty"`
	Description string `json:"description"`
	MemberCount int64  `json:"member_count"`
	ChildCount  int64  `json:"child_count"`
	// GrantedCount counts grant references (skill) — deletion guard.
	GrantedCount int64 `json:"granted_count"`
}

// ListDepartments returns every department with admin-view fields.
func ListDepartments(db *sql.DB) ([]DepartmentInfo, error) {
	rows, err := db.Query(`SELECT g.id, g.name, g.parent_id, g.leader_id, g.description,
		COALESCE(u.username, ''),
		(SELECT COUNT(*) FROM user_groups ug WHERE ug.group_id = g.id),
		(SELECT COUNT(*) FROM groups c WHERE c.parent_id = g.id),
		(SELECT COUNT(*) FROM app_grants sg WHERE sg.grantee_type = 'group' AND ` + ciColumnCmp("sg.grantee", "g.name") + `)
		FROM groups g LEFT JOIN users u ON u.id = g.leader_id
		ORDER BY g.id`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []DepartmentInfo
	for rows.Next() {
		var d DepartmentInfo
		if err := rows.Scan(&d.ID, &d.Name, &d.ParentID, &d.LeaderID, &d.Description,
			&d.LeaderName, &d.MemberCount, &d.ChildCount, &d.GrantedCount); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

// GroupByID returns one group.
func GroupByID(db *sql.DB, id int64) (*Group, error) {
	row := db.QueryRow(`SELECT id, name, parent_id, leader_id, description FROM groups WHERE id = ?`, id)
	var g Group
	err := row.Scan(&g.ID, &g.Name, &g.ParentID, &g.LeaderID, &g.Description)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, ErrNotFound
	}
	return &g, err
}

// CreateDepartment inserts a department (parent_id 0 = top level).
// The parent must exist; duplicate names are rejected (UNIQUE, NOCASE since 0019).
// leader 必须存在(与 UpdateDepartment 同口径,防悬空主管静默落库)。
func CreateDepartment(db *sql.DB, name string, parentID, leaderID int64, description string) (int64, error) {
	if name == EveryoneGroupName {
		return 0, ErrValidation // 保留名:隐式全员组不可重建/改名占用
	}
	if name == "" {
		return 0, ErrValidation
	}
	if parentID != 0 {
		if _, err := GroupByID(db, parentID); err != nil {
			return 0, err
		}
	}
	if leaderID != 0 {
		if _, err := GetUserByID(db, leaderID); err != nil {
			return 0, err
		}
	}
	id, err := InsertID(db, `INSERT INTO groups (name, parent_id, leader_id, description) VALUES (?, ?, ?, ?)`,
		name, parentID, leaderID, description)
	if err != nil {
		if isUniqueViolation(err) {
			return 0, ErrDuplicate
		}
		return 0, err
	}
	InvalidateGroupTree() // F3: 新部门立即进入组织树(环检测/授权/预算链)
	return id, nil
}

// UpdateDepartment renames/reparents/re-leads a department.
// Guards: parent must exist and not be the department itself or a
// descendant (cycle); leader must exist. Renames cascade to the grant
// tables so existing grants keep resolving (授权按组名,改名不得静默失效).
//
// 2026-09-11:部门预算(budget_money)下线(钱只由余额闸门管),本函数不再带预算
// 参数;数据库列保留但不再读写,避免破坏性迁移。
func UpdateDepartment(db *sql.DB, id int64, name string, parentID, leaderID int64, description string) error {
	g, err := GroupByID(db, id)
	if err != nil {
		return err
	}
	if parentID != 0 {
		if parentID == id {
			return ErrValidation
		}
		sub, err := subtreeGroupIDs(db, id)
		if err != nil {
			return err
		}
		for _, s := range sub {
			if s == parentID {
				return ErrValidation
			}
		}
		if _, err := GroupByID(db, parentID); err != nil {
			return err
		}
	}
	if leaderID != 0 {
		if _, err := GetUserByID(db, leaderID); err != nil {
			return err
		}
	}
	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	// 保留名:全员行不可改名(隐式全员授权按名解析,改名会静默失效;
	// 迁移 0018 一次性 seed,无启动自愈)。
	if g.Name == EveryoneGroupName {
		return ErrValidation
	}
	if name == EveryoneGroupName {
		return ErrValidation // 保留名
	}
	if name == "" {
		return ErrValidation // 空名会把全部授权级联写成空串,store 层兜底拒绝
	}
	if name != "" && name != g.Name {
		// COLLATE NOCASE 与授权解析一致:授权存储大小写异于组名的(手输/LDAP)改名后不失效
		if _, err := tx.Exec("UPDATE app_grants SET grantee = ? WHERE grantee_type = 'group' AND "+CaseInsensitiveCmp("grantee"), name, g.Name); err != nil {
			return err
		}
		// 审计修复 2026-P (H1): 0036 共享资源授权表同样级联改名,否则被授权
		// 部门改名后共享技能/Agent 授权静默失效(陈旧 grantee)。
		// P2:授权已统一到 app_grants,上面的 UPDATE 已覆盖全部能力类型。
	}
	_, err = tx.Exec(`UPDATE groups SET name = ?, parent_id = ?, leader_id = ?, description = ? WHERE id = ?`,
		name, parentID, leaderID, description, id)
	if err != nil {
		if isUniqueViolation(err) {
			return ErrDuplicate
		}
		return err
	}
	if err := tx.Commit(); err != nil {
		return err
	}
	InvalidateGroupTree() // F3: 改名/改父级/换主管立即生效
	return nil
}

// DeleteDepartment removes a department. Guard: departments with members,
// child departments or grant references cannot be deleted — the operator
// must move members/children and clear grants first (避免误删权限黑洞).
// 守卫计数与删除同事务(TOCTOU);授权引用按 NOCASE 计数(与解析口径一致,
// 大小写变体的授权不得绕过守卫导致孤儿授权/后续复活)。
func DeleteDepartment(db *sql.DB, id int64) error {
	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if g, err := GroupByID(db, id); err == nil && g.Name == EveryoneGroupName {
		return ErrValidation // 保留名
	}
	// P2 后三张授权表合并为 app_grants:此处从「数三张表」收敛为一次统计
	// (此前第三项仍落在旧表 agent_preset_grants 上,是三表合并时的漏改)。
	var memberCount, childCount, grantCount int64
	if err := tx.QueryRow(`SELECT
		(SELECT COUNT(*) FROM user_groups ug WHERE ug.group_id = g.id),
		(SELECT COUNT(*) FROM groups c WHERE c.parent_id = g.id),
		(SELECT COUNT(*) FROM app_grants ag WHERE ag.grantee_type = 'group' AND `+ciColumnCmp("ag.grantee", "g.name")+`)
		FROM groups g WHERE g.id = ?`, id).Scan(&memberCount, &childCount, &grantCount); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return ErrNotFound
		}
		return err
	}
	if memberCount > 0 || childCount > 0 || grantCount > 0 {
		return ErrDepartmentInUse
	}
	if _, err := tx.Exec("DELETE FROM groups WHERE id = ?", id); err != nil {
		return err
	}
	if err := tx.Commit(); err != nil {
		return err
	}
	InvalidateGroupTree() // F3: 删除后立刻从树中移除
	return nil
}

// subtreeGroupIDs returns the department id plus all descendant ids
// (in-memory walk; the tree is small). 复用 loadGroupTree 的缓存:
// 不再每次全表 SELECT groups(与 EffectiveDeptBudget 共用同一缓存树,
// 变更后应失效组织树缓存(当前依赖 TTL 兜底)。
func subtreeGroupIDs(db *sql.DB, rootID int64) ([]int64, error) {
	nodes, err := loadGroupTree(db)
	if err != nil {
		return nil, err
	}
	children := map[int64][]int64{}
	for _, n := range nodes {
		children[n.parent] = append(children[n.parent], n.id)
	}
	// visited 兜底(F3):迁移/手工数据可能留下环,绝不让配额热路径死循环。
	seen := map[int64]bool{rootID: true}
	out := []int64{rootID}
	stack := []int64{rootID}
	for len(stack) > 0 {
		cur := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		for _, c := range children[cur] {
			if seen[c] {
				continue
			}
			seen[c] = true
			out = append(out, c)
			stack = append(stack, c)
		}
	}
	return out, nil
}

File: rust/crates/dsh-store/src/departments.rs
//! 部门域（Go `serverstore/departments.go` 等价）。

use crate::errors::{map_db_error, StoreError};
use sqlx::Row;

/// 隐式全员组名（所有用户隐式成员，授权给全员=授权给所有人）。
pub const EVERYONE_GROUP_NAME: &str = "全员";

/// Group 组织单元：部门（可选 parent 金字塔 + leader 主管）。
#[derive(Debug, Clone, Default)]
pub struct Group {
    pub id: i64,
    pub name: String,
    pub parent_id: i64,
    pub leader_id: i64,
    pub description: String,
}

/// DepartmentInfo 管理端视图的部门形状。
#[derive(Debug, Clone, Default)]
pub struct DepartmentInfo {
    pub id: i64,
    pub name: String,
    pub parent_id: i64,
    pub leader_id: i64,
    pub leader_name: String,
    pub description: String,
    pub member_count: i64,
    pub child_count: i64,
    pub granted_count: i64,
    pub budget_money: Option<f64>,
    pub monthly_cost: f64,
}

/// ListDepartments 返回全部部门（含管理端字段）。
pub async fn list_departments(pool: &sqlx::PgPool) -> Result<Vec<DepartmentInfo>, StoreError> {
    let rows = sqlx::query(
        r#"SELECT g.id::bigint, g.name, g.parent_id::bigint, g.leader_id::bigint, g.description,
        COALESCE(u.username, ''),
        (SELECT COUNT(*) FROM user_groups ug WHERE ug.group_id = g.id),
        (SELECT COUNT(*) FROM groups c WHERE c.parent_id = g.id),
        (SELECT COUNT(*) FROM app_grants sg WHERE sg.grantee_type = 'group' AND LOWER(sg.grantee) = LOWER(g.name)),
        g.budget_money
        FROM groups g LEFT JOIN users u ON u.id = g.leader_id
        ORDER BY g.id"#,
    )
    .fetch_all(pool)
    .await
    .map_err(map_db_error)?;
    let mut out = Vec::new();
    for r in rows {
        let budget_money: Option<f64> = r.get("budget_money");
        let leader_id: i64 = r.get("leader_id");
        let leader_name: String = r.get("leader_name");
        out.push(DepartmentInfo {
            id: r.get("id"),
            name: r.get("name"),
            parent_id: r.get("parent_id"),
            leader_id,
            leader_name,
            description: r.get("description"),
            member_count: r.get("member_count"),
            child_count: r.get("child_count"),
            granted_count: r.get("granted_count"),
            budget_money,
            monthly_cost: 0.0, // 由调用方 DeptMonthlyCostBatch 批量填充
        });
    }
    Ok(out)
}

/// GroupByID 返回单个部门。
pub async fn group_by_id(pool: &sqlx::PgPool, id: i64) -> Result<Group, StoreError> {
    let row = sqlx::query("SELECT id, name, parent_id, leader_id, description FROM groups WHERE id = $1")
        .bind(id)
        .fetch_optional(pool)
        .await
        .map_err(map_db_error)?
        .ok_or(StoreError::NotFound)?;
    Ok(Group {
        id: row.get::<i64, _>("id"),
        name: row.get("name"),
        parent_id: row.get::<i32, _>("parent_id") as i64,
        leader_id: row.get::<i32, _>("leader_id") as i64,
        description: row.get("description"),
    })
}

/// CreateDepartment 插入部门（parent_id 0 = 顶级）。
pub async fn create_department(
    pool: &sqlx::PgPool,
    name: &str,
    parent_id: i64,
    leader_id: i64,
    description: &str,
) -> Result<i64, StoreError> {
    if name == EVERYONE_GROUP_NAME || name.is_empty() {
        return Err(StoreError::Validation);
    }
    if parent_id != 0 {
        group_by_id(pool, parent_id).await?;
    }
    if leader_id != 0 {
        crate::users::get_user_by_id(pool, leader_id).await?;
    }
    let row = sqlx::query("INSERT INTO groups (name, parent_id, leader_id, description) VALUES ($1, $2, $3, $4) RETURNING id")
        .bind(name)
        .bind(parent_id)
        .bind(leader_id)
        .bind(description)
        .fetch_one(pool)
        .await
        .map_err(map_db_error)?;
    Ok(row.get("id"))
}

/// UpdateDepartment 改名/改上级/改主管。守卫：上级必须存在且非自身/后代（防环）、主管存在。
pub async fn update_department(
    pool: &sqlx::PgPool,
    id: i64,
    name: &str,
    parent_id: i64,
    leader_id: i64,
    description: &str,
) -> Result<(), StoreError> {
    update_department_with_budget_inner(pool, id, name, parent_id, leader_id, description, None).await
}

/// UpdateDepartmentWithBudget 同语义，额外同一事务内设置预算（nil 不变，0 清除，>0 设置）。
pub async fn update_department_with_budget(
    pool: &sqlx::PgPool,
    id: i64,
    name: &str,
    parent_id: i64,
    leader_id: i64,
    description: &str,
    budget: Option<f64>,
) -> Result<(), StoreError> {
    update_department_with_budget_inner(pool, id, name, parent_id, leader_id, description, budget).await
}

async fn update_department_with_budget_inner(
    pool: &sqlx::PgPool,
    id: i64,
    name: &str,
    parent_id: i64,
    leader_id: i64,
    description: &str,
    budget: Option<f64>,
) -> Result<(), StoreError> {
    let g = group_by_id(pool, id).await?;
    if parent_id != 0 {
        if parent_id == id {
            return Err(StoreError::Validation);
        }
        let sub = subtree_group_ids(pool, id).await?;
        if sub.contains(&parent_id) {
            return Err(StoreError::Validation);
        }
        group_by_id(pool, parent_id).await?;
    }
    if leader_id != 0 {
        crate::users::get_user_by_id(pool, leader_id).await?;
    }
    let mut tx = pool.begin().await.map_err(map_db_error)?;
    if g.name == EVERYONE_GROUP_NAME || name == EVERYONE_GROUP_NAME || name.is_empty() {
        return Err(StoreError::Validation);
    }
    if name != g.name {
        // 改名级联授权表（大小写不敏感）
        sqlx::query("UPDATE app_grants SET grantee = $1 WHERE grantee_type = 'group' AND LOWER(grantee) = LOWER($2)")
            .bind(name)
            .bind(&g.name)
            .execute(&mut *tx)
            .await
            .map_err(map_db_error)?;
    }
    let res = sqlx::query("UPDATE groups SET name = $1, parent_id = $2, leader_id = $3, description = $4 WHERE id = $5")
        .bind(name)
        .bind(parent_id)
        .bind(leader_id)
        .bind(description)
        .bind(id)
        .execute(&mut *tx)
        .await
        .map_err(map_db_error)?;
    if res.rows_affected() == 0 {
        return Err(StoreError::NotFound);
    }
    if let Some(b) = budget {
        if b < 0.0 {
            return Err(StoreError::Validation);
        }
        if b == 0.0 {
            sqlx::query("UPDATE groups SET budget_money = NULL WHERE id = $1")
                .bind(id)
                .execute(&mut *tx)
                .await
                .map_err(map_db_error)?;
        } else {
            sqlx::query("UPDATE groups SET budget_money = $1 WHERE id = $2")
                .bind(b)
                .bind(id)
                .execute(&mut *tx)
                .await
                .map_err(map_db_error)?;
        }
    }
    tx.commit().await.map_err(map_db_error)?;
    Ok(())
}

/// DeleteDepartment 删除部门。守卫：有成员/子部门/授权引用不可删（TOCTOU 内同事务计数）。
pub async fn delete_department(pool: &sqlx::PgPool, id: i64) -> Result<(), StoreError> {
    let mut tx = pool.begin().await.map_err(map_db_error)?;
    if let Ok(g) = crate::departments::group_by_id(pool, id).await {
        if g.name == EVERYONE_GROUP_NAME {
            return Err(StoreError::Validation);
        }
    } else {
        return Err(StoreError::NotFound);
    }
    let row = sqlx::query(
        "SELECT
        (SELECT COUNT(*) FROM user_groups ug WHERE ug.group_id = g.id),
        (SELECT COUNT(*) FROM groups c WHERE c.parent_id = g.id),
        (SELECT COUNT(*) FROM app_grants ag WHERE ag.grantee_type = 'group' AND LOWER(ag.grantee) = LOWER(g.name))
        FROM groups g WHERE g.id = $1",
    )
    .bind(id)
    .fetch_optional(&mut *tx)
    .await
    .map_err(map_db_error)?
    .ok_or(StoreError::NotFound)?;
    let member_count: i64 = row.get(0);
    let child_count: i64 = row.get(1);
    let grant_count: i64 = row.get(2);
    if member_count > 0 || child_count > 0 || grant_count > 0 {
        return Err(StoreError::DepartmentInUse);
    }
    sqlx::query("DELETE FROM groups WHERE id = $1")
        .bind(id)
        .execute(&mut *tx)
        .await
        .map_err(map_db_error)?;
    tx.commit().await.map_err(map_db_error)?;
    Ok(())
}

/// subtree_group_ids 返回部门 id 及其全部后代 id（内存遍历，树很小）。
pub async fn subtree_group_ids(pool: &sqlx::PgPool, root_id: i64) -> Result<Vec<i64>, StoreError> {
    let nodes: Vec<(i64, i32)> = sqlx::query_as("SELECT id, parent_id FROM groups")
        .fetch_all(pool)
        .await
        .map_err(map_db_error)?;
    let mut children: std::collections::HashMap<i64, Vec<i64>> = std::collections::HashMap::new();
    for (id, parent) in nodes {
        children.entry(parent as i64).or_default().push(id);
    }
    let mut out = vec![root_id];
    let mut stack = vec![root_id];
    while let Some(cur) = stack.pop() {
        if let Some(cs) = children.get(&cur) {
            for c in cs {
                out.push(*c);
                stack.push(*c);
            }
        }
    }
    Ok(out)
}
